### ISSUE TITLE
Possible feature: wrap apply functions to enable callbacks (especially for saving trajectories)

### ISSUE THREAD

**Issue by cpgoodri** [google/jax-md#172]
Saving trajectories is obviously a common use case. Currently, there are a few ways of doing it. For example, I often use something like:
```python
def apply_single_step(state, t):
    return apply(state, t=t), 0
def apply_many_steps(state, t_list):
    state, _ = lax.scan(apply_single_step, state, t_list)
    return state, state.position

final_state, trajectory = lax.scan(
    apply_many_steps,
    initial_state,
    jnp.arange(num_total_steps).reshape(
        num_total_steps // record_every, record_every
    ),
)
```
However, this requires saving the entire (saved) trajectory in memory, and isn't ideal when the `record_every` interval isn't commensurate with other reasons intervals (e.g. neighbor_list buffer checks, gradient checkpoints, other measurements). xarray and the underlying netcdf file format have the ideal capabilities for saving trajectories, but unfortunately are not jax-compatible. I'm wondering if there is a way to wrap the apply function such that at specific intervals it launches a callback function that saves data the state. Since saving doesn't affect the computation, my understanding is that the simulation could continue executing on the device without waiting for the callback function to finish, meaning this should in principle have very little effect on speed. However, I have zero experience with jax callbacks so this could be a terrible idea.

Perhaps more generally, would it be possible/useful for jaxmd to have the ability to wrap apply functions with user-defined callbacks? In addition to saving data, this would be useful for real-time data analysis that includes non-jax libraries.

**Comment by sschoenholz** [comment]
Hey Carl,

Thanks for bringing this up! I think this feature could be interesting, but I'm not sure I totally understand. Can you write out the semantics that you're hoping for? At the moment, what you've described would be very do-able, but I'm not sure I follow what part should be built into jax md?

For example, it sort of seems like adding a call to the `id_tap` function from [jax.experimental.host_callback](https://github.com/google/jax/blob/main/jax/experimental/host_callback.py) would allow you to write code to save states via a callback without modifying jax md. 

I guess one piece of functionality that jax md could provide is a higher level function function that is sort of like `jax_md.run_simulation` which would run many steps of simulation and provide some nice utilities for checkpointing, running callbacks every T steps, etc... Is this sort of what you have in mind?

**Comment by cpgoodri** [comment]
Hummm... you might be right that `id_tap` is all you need and this is better kept outside of jaxmd. As I said, I don't have experience with callbacks and I'm still learning how they work. I do think some sort of high-level utility function might be nice (especially since the callback API is still experimental) but only if it would actually be useful. Perhaps a middle ground would be a really short tutorial showing how to use the `id_tap` to save a trajectory or some scipy-based measurement? I/my group would be happy to put this together if/when we get around to doing it (our current need is not yet urgent).
